lib/src/winrt/internal/async_helpers.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import '../../exceptions.dart';
+import '../foundation/enums.g.dart';
+import '../foundation/iasyncaction.dart';
+import '../foundation/iasyncinfo.dart';
+import '../foundation/iasyncoperation.dart';
+
+/// Completes the given [completer] when [asyncAction] completes.
+void completeAsyncAction(
+  IAsyncAction asyncAction,
+  Completer<void> completer,
+) =>
+    _completeAsyncDelegate(asyncAction, completer, completer.complete);
+
+/// Completes the given [completer] when [asyncOperation] completes.
+///
+/// If [asyncOperation] completes successfully, [completer] will be completed
+/// with the [value].
+void completeAsyncOperation<T, C>(
+  IAsyncOperation<T> asyncOperation,
+  Completer<C> completer,
+  C Function() value,
+) =>
+    _completeAsyncDelegate(
+        asyncOperation, completer, () => completer.complete(value()));
+
+/// Completes the given [completer] by polling [asyncDelegate]'s `status`
+/// property every `10` milliseconds until the [asyncDelegate] completes.
+///
+/// [onCompleted] is called when the [asyncDelegate] completes successfully.
+Future<void> _completeAsyncDelegate<T extends IAsyncInfo, C>(T asyncDelegate,
+    Completer<C> completer, void Function() onCompleted) async {
+  try {
+    while (asyncDelegate.status == AsyncStatus.started) {
+      // Yield execution to Dart for 10 milliseconds to avoid blocking the app.
+      await Future<void>.delayed(const Duration(milliseconds: 10));
+    }
+
+    switch (asyncDelegate.status) {
+      // This case is handled by the while loop above.
+      case AsyncStatus.started:
+        break;
+      case AsyncStatus.completed:
+        onCompleted();
+        break;
+      case AsyncStatus.canceled:
+        completer.completeError('The async operation canceled!');
+        break;
+      case AsyncStatus.error:
+        completer.completeError(WindowsException(asyncDelegate.errorCode));
```

Comment:
I think asyncDelegate.errorCode is an `HSTRING`, not an `HRESULT`, no? Based on this: https://learn.microsoft.com/en-us/uwp/api/windows.foundation.iasyncinfo.errorcode?view=winrt-22621#windows-foundation-iasyncinfo-errorcode

WindowsException takes an `HRESULT` and uses `FormatMessage` to convert it to a String.

Reply:
I'm pretty sure this returns an `HRESULT`. C# wraps this with `System.Exception`. If you change the language to `C++/WinRT`, you will see that it returns an HRESULT.
![image](https://user-images.githubusercontent.com/45524029/210129713-baa1b1aa-534a-4d6a-a21a-aec2280d95d7.png)

I also checked the `asyncinfo.h` file from the Windows SDK:
```h
 virtual /* [propget] */ HRESULT STDMETHODCALLTYPE get_ErrorCode( 
            /* [retval][out] */ __RPC__out HRESULT *errorCode) = 0;
```

Reply:
Doh! You're totally right. It makes sense that the Microsoft language projections wrap various low-level concepts, but it's so confusing that their docs then tell lies about the interface signatures! I want a drop-down menu choice for "tell me the truth, not some language-specific warped reality" :)